src/v/rpc/parse_utils.h

```diff
@@ -275,25 +276,25 @@ ss::future<T> parse_type(ss::input_stream<char>& in, const header& h) {
     return read_iobuf_exactly(in, h.payload_size).then([h](iobuf io) {
         validate_payload_and_header(io, h);
 
-        switch (h.compression) {
-        case compression_type::none:
-            break;
-
-        case compression_type::zstd: {
-            compression::stream_zstd fn;
-            io = fn.uncompress(std::move(io));
-            break;
-        }
-
-        default:
-            return ss::make_exception_future<T>(std::runtime_error(
+        auto iobuf_fut = ss::do_with(std::move(io), [h](iobuf& io) {
+            switch (h.compression) {
+            case compression_type::none:
+                return ss::make_ready_future<iobuf>(std::move(io));
+            case compression_type::zstd: {
+                compression::async_stream_zstd fn;
+                return fn.uncompress(io);
+            }
+            }
+            return ss::make_exception_future<iobuf>(std::runtime_error(
               fmt::format("no compression supported. header: {}", h)));
-        }
+        });
```

Comment:
Does the check on `h.compression` need to be inside the async part?

Seems like you could do the conditional up front, and then only go into creating futures for compression if compression is in use.  I'm doubtful that the compiler will be able to do this optimization itself.  It'll only save a few cycles, but this is a very hot path so likely to be worthwhile.